crates/rust-analyzer/src/config.rs

```diff
@@ -760,8 +766,49 @@ fn schema(fields: &[(&'static str, &'static str, &[&str], &str)]) -> serde_json:
     map.into()
 }
 
+enum DocMode {
+    Json,
+    #[cfg(test)]
+    AsciiDoc,
+}
+
+fn generate_doc<'a>(doc: &'a [&str], mode: DocMode) -> String {
+    // asciidoctor syntax looks like
+    // ```
+    // some docs here, with a new line +
+    // and docs following
+    // ```
+    // Collapse regular newlines which are probably just word-wrapped, but leave entirely empty lines as hard breaks.
+    use std::borrow::Cow;
+    let line_ending = if matches!(mode, DocMode::Json) { "\n" } else { " +\n" };
+    let doc_line = |line: &'a str| -> &'a str {
+        if matches!(mode, DocMode::Json) {
+            line.trim_start()
+        } else {
+            line
+        }
+    };
+    doc.windows(2)
+        .map(|window| {
+            let doc_line = doc_line(window[0]);
+            // Add `+` if this is asciidoctor syntax and should be a hard break
+            if window[1].is_empty() && !doc_line.is_empty() {
+                Cow::Owned(format!("{}{}", doc_line, line_ending))
+            } else {
+                doc_line.into()
+            }
+        })
+        // Remove the empty line
+        .filter(|line| !line.is_empty())
+        // `.windows()` doesn't include the last line
+        .chain(doc.last().map(|&x| doc_line(x).into()))
+        // When emitting JSON, we trim the start of the line, so we need to add back a space between
+        // word-wrapped lines.
+        .join(if matches!(mode, DocMode::Json) { " " } else { "" })
```

Comment:
This is not 100% correct, it inserts a space after hard breaks: https://github.com/rust-analyzer/rust-analyzer/blob/6ccd45fdb9d458bda4941d8a6575c8407f00431d/editors/code/package.json#L662. But fixing it seemed complicated and I'm not sure how much it affects in practice.